Assets/MRTK/SDK/Experimental/Features/UX/BoundsControl/Visuals/TranslationHandles.cs

```diff
@@ -0,0 +1,338 @@
+                    // attach new collider
+                    var handleBounds = VisualUtils.GetMaxBounds(GetVisual(handle).gameObject);
+                    var invScale = handleBounds.size.x == 0.0f ? 0.0f : config.HandleSize / handleBounds.size.x;
+                    Vector3 colliderSizeScaled = handleBounds.size * invScale;
+                    Vector3 colliderCenterScaled = handleBounds.center * invScale;
+                    if (config.TranslationHandlePrefabColliderType == HandlePrefabCollider.Box)
+                    {
+                        BoxCollider collider = handle.gameObject.AddComponent<BoxCollider>();
+                        collider.size = colliderSizeScaled;
+                        collider.center = colliderCenterScaled;
+                        collider.size += config.ColliderPadding;
+                    }
+                    else
+                    {
+                        SphereCollider sphere = handle.gameObject.AddComponent<SphereCollider>();
+                        sphere.center = colliderCenterScaled;
+                        sphere.radius = colliderSizeScaled.x * 0.5f;
+                        sphere.radius += VisualUtils.GetMaxComponent(config.ColliderPadding);
+                    }
+                }
+
+            }
+        }
+
+        internal const int NumFaces = 6;
+
+        private Vector3[] faceCenters = new Vector3[NumFaces];
+        private CardinalAxisType[] faceAxes;
+
+        internal int GetTranslationHandleIdx(Transform handle)
+        {
+            for (int i = 0; i < handles.Count; ++i)
+            {
+                if (handle == handles[i])
+                {
+                    return i;
+                }
+            }
+
+            return handles.Count;
+        }
+
+        internal Vector3 GetFaceCenter(int index)
+        {
+            Debug.Assert(index >= 0 && index <= NumFaces, "Face center index out of bounds");
+            return faceCenters[index];
+        }
+
+        internal CardinalAxisType GetAxisType(int index)
+        {
+            Debug.Assert(index >= 0 && index <= NumFaces, "Face axes index out of bounds");
+            return faceAxes[index];
+        }
+
+        internal CardinalAxisType GetAxisType(Transform handle)
+        {
+            int index = GetTranslationHandleIdx(handle);
+            return GetAxisType(index);
+        }
+
+        private void UpdateHandles()
+        {
+            for (int i = 0; i < handles.Count; ++i)
+            {
+                handles[i].position = GetFaceCenter(i);
+            }
+        }
+
+        internal void CalculateFaceCenters(ref Vector3[] boundsCorners)
+        {
+            if (boundsCorners != null && faceCenters != null)
+            {
+                for (int i = 0; i < faceCenters.Length; ++i)
+                {
+                    faceCenters[i] = VisualUtils.GetFaceCenterPosition(i, ref boundsCorners);
+                }
+            }
+
+            UpdateHandles();
```

Comment:
Nit: Should this go into the if statement above:

if (boundsCorners != null && faceCenters != null)

Or is there some case where these are both null that UpdateHandles will succeed?

Reply:
Not sure. This is the way that `RotationHandles` is written. @thalbern could you please elucidate the logic behind this ordering?

Reply:
i think it's fine to have this inside the if - fingers crossed - if not we're missing a test to catch this - but looks good now that i moved it in :)